env::split_paths regression in nightly-2025-08-31

This code used to compile in previous stable releases and nightlies prior to nightly-2025-08-31. After https://github.com/rust-lang/rust/pull/145242, it no longer compiles, which seems unintended.

```rust
pub fn repro() -> Option<std::path::PathBuf> {
    let unparsed = std::ffi::OsString::new();
    std::env::split_paths(&unparsed).next()
}
```

`rustc --crate-type=lib --edition=2021 --target=x86_64-unknown-linux-gnu repro.rs`

```console
error[E0597]: `unparsed` does not live long enough
 --> repro.rs:3:27
  |
2 |     let unparsed = std::ffi::OsString::new();
  |         -------- binding `unparsed` declared here
3 |     std::env::split_paths(&unparsed).next()
  |     ----------------------^^^^^^^^^-
  |     |                     |
  |     |                     borrowed value does not live long enough
  |     a temporary with access to the borrow is created here ...
4 | }
  | -
  | |
  | `unparsed` dropped here while still borrowed
  | ... and the borrow might be used here, when that temporary is dropped and runs the destructor for type `SplitPaths<'_>`
  |
  = note: the temporary is part of an expression at the end of a block;
          consider forcing this temporary to be dropped sooner, before the block's local variables are dropped
help: for example, you could save the expression's value in a new local variable `x` and then make `x` be the expression at the end of the block
  |
3 |     let x = std::env::split_paths(&unparsed).next(); x
  |     +++++++                                        +++
```

The regression only occurs if building for **both** a unix target platforms **and** an edition older than 2024. It does not reproduce in non-unix platforms (`--target=x86_64-pc-windows-mscv`) nor 2024 edition (`--edition=2024`).

@joboet

Releases of `cc` from 1.0.70 (August 31, 2021) through 1.0.100 (June 23, 2024) are affected. (These are the releases containing https://github.com/rust-lang/cc-rs/pull/612 but not https://github.com/rust-lang/cc-rs/pull/1103.)

```console
error[E0597]: `path_entries` does not live long enough
    --> $CARGO_HOME/registry/src/index.crates.io-1949cf8c6b5b557f/cc-1.0.100/src/lib.rs:4134:22
     |
4133 |     let path_entries = path_entries.or(env::var_os("PATH"))?;
     |         ------------ binding `path_entries` declared here
4134 |     env::split_paths(&path_entries).find_map(|path_entry| {
     |     -----------------^^^^^^^^^^^^^-
     |     |                |
     |     |                borrowed value does not live long enough
     |     a temporary with access to the borrow is created here ...
...
4142 | }
     | -
     | |
     | `path_entries` dropped here while still borrowed
     | ... and the borrow might be used here, when that temporary is dropped and runs the destructor for type `SplitPaths<'_>`
     |
     = note: the temporary is part of an expression at the end of a block;
             consider forcing this temporary to be dropped sooner, before the block's local variables are dropped
help: for example, you could save the expression's value in a new local variable `x` and then make `x` be the expression at the end of the block
     |
4134 ~     let x = env::split_paths(&path_entries).find_map(|path_entry| {
4135 |         let mut exe = path_entry.join(tool);
 ...
4140 |         }
4141 ~     }); x
     |
```

--- a/library/std/src/sys/pal/unix/os.rs
+++ b/library/std/src/sys/pal/unix/os.rs
@@ -186,14 +186,24 @@ pub fn chdir(p: &path::Path) -> io::Result<()> {
     if result == 0 { Ok(()) } else { Err(io::Error::last_os_error()) }
 }
 
-pub type SplitPaths<'a> = impl Iterator<Item = PathBuf>;
+// This can't just be `impl Iterator` because that requires `'a` to be live on
+// drop (see #146045).
+pub type SplitPaths<'a> = iter::Map<
+    slice::Split<'a, u8, impl FnMut(&u8) -> bool + 'static>,
+    impl FnMut(&[u8]) -> PathBuf + 'static,
+>;
 
 #[define_opaque(SplitPaths)]
 pub fn split_paths(unparsed: &OsStr) -> SplitPaths<'_> {
-    unparsed
-        .as_bytes()
-        .split(|&b| b == PATH_SEPARATOR)
-        .map(|part| PathBuf::from(OsStr::from_bytes(part)))
+    fn is_separator(&b: &u8) -> bool {
+        b == PATH_SEPARATOR
+    }
+
+    fn into_pathbuf(part: &[u8]) -> PathBuf {
+        PathBuf::from(OsStr::from_bytes(part))
+    }
+
+    unparsed.as_bytes().split(is_separator).map(into_pathbuf)
 }
 
 #[derive(Debug)]

--- a/tests/ui/type-alias-impl-trait/split-paths-may-dangle.rs
+++ b/tests/ui/type-alias-impl-trait/split-paths-may-dangle.rs
@@ -1,0 +1,11 @@
+// Regression test for issue #146045 - ensure that the TAIT `SplitPaths` does not
+// require the borrowed string to be live.
+//@ check-pass
+//@ edition:2015
+
+pub fn repro() -> Option<std::path::PathBuf> {
+    let unparsed = std::ffi::OsString::new();
+    std::env::split_paths(&unparsed).next()
+}
+
+fn main() {}

